Kckoh

kckoh opened the pull request:
Implement GET / Tasks

아직 get  테스트 통과 하지 못했습니다. npm run test를 돌렸을때 한글이 깨지는데 어떻게하면 해결 할 수 있을까요?

kckoh committed:
Implement GET /tasks

코드 리뷰 부탁드립니다. npm run test 실행 했을때 한글이 인식이 안되는데 어떻게 하면 될까요? 다른 부분들도 리뷰 해주시면 감사하겠습니다.

kckoh committed:
Implemente GET /

아직 GET 부분을 하고 있는데 막히고 있네요. npm run test했을때 한글이 인식이 안되는데 어떻게 하면 될까요?

ssisksl77 commented:
`npm run test` 를 돌리면 한글이 깨진다는 것이 무엇일까요?! 메시지들이 깨진다는 말씀이실까요?

<img width="551" alt="image" src="https://user-images.githubusercontent.com/15942102/190182047-edf59c2f-ef08-4c26-aeac-13e96a90e53c.png">

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,114 @@
+
+    private void getMappingTasks(String method, String path) throws IOException {
+
+//        GET /tasks
+        if (method.equals("GET") && path.equals("/tasks")) {
+            content = tasksToJson();
+            return;
+        }
+//        GET /tasks/{id}
+        else if (method.equals("GET") && path.contains("/tasks/")) {
+            String[] split = path.split("/");
+
+            String id = split[2];
+            if (!isNumeric(id)) {
+                System.out.println("tasks id not a number");
+                status = 404;
+                return;
+            }
+
+            for (Task task : tasks) {
+                if (task.getId() == Long.parseLong(id)) {
+                    content = taskToJson(task);
+                    status = 200;
+                    return;
+                }
+            }
+
+            content = "";
+        }
+
+
```
좋아요! 우선 중첩된 if문을 없애는 연습을 해보는 것이 어떨까요?!

중첩된 if문을 가독성을 저해합니다. 하지만 일단 코드가 돌아가도록 만들고 나중에 해도 됩니다!

kckoh replied:
그러면 if문을 따로 extract method을 사용해서 빼는게 좋을까요? 그러면 너무 코드가 길어질거 같아서 고민이 되네요

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+
+    public static boolean isNumeric(String str) {
```
아주 일반적인 숫자를 검사하는 코드군요! 

문자열에 관해서는 아파치에서 만들어놓은 최적화된 라이브러리를 써보는 것도 나쁘지 않아요!

https://commons.apache.org/ 

한번 이것을 보고 대체해보는 것은 어떨까요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
```
벌써 롬복까지...!!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,114 @@
+            tasks.add(task1);
+        }
+
+        getMappingTasks(method, path);
+
+
+        if (method.equals("POST") && path.equals("/tasks")) {
+            content = "Create a new task";
+        }
```
물론 첫술에 배부를 수는 없지만 
현재는 getMapping에서 매핑이 성공되어도
불필요하게 POST 매핑을 검사하는 것 같군요!

kckoh replied:
그러면 스위치문으로 짜서 하는게 좋은 방법 일까요?

ssisksl77 replied:
그것도 하나의 방법일 수 있겠네요! 
일단 여기서 중요한 것은 어느 한 매핑이 성공하면 다른 매핑에 검사는 아예 할 필요가 없도록 만들어보는 것이 먼저입니다!

아름답게 만드는 코드는 그 다음에 생각해봐요 우리!

kckoh commented:
[](url)
<img width="960" alt="error" src="https://user-images.githubusercontent.com/43446639/190264782-86d03f14-858e-4693-8242-00774816ec9f.png">

kckoh commented:
> <img alt="error" width="960" src="https://user-images.githubusercontent.com/43446639/190264782-86d03f14-858e-4693-8242-00774816ec9f.png">

지금 windows 환경에서 gitbash를 쓰고 있는데 어디서 문제 인지 모르겠네요.

kckoh committed:
Refactor some code based on reviews

리뷰 부탁드립니다.
고친점:
    if문은 전부 따로 다른 method으로 분리
    아파치 library 사용
    get/post/delete/post/put 중복 X
    HTTP STATUS Code/ HTTP METHOD enum 클라스로 만들기

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/enums/HttpStatus.java:
```diff
@@ -0,0 +1,16 @@
+
+
+@Getter
+@AllArgsConstructor
```
이 코드에서 문제가 없을 수 있겠지만! 

lombok의 `@AllArgsConstructor` 를 주의하라는 글이 있습니다. 

슬쩍 공유드려요~

https://medium.com/geekculture/good-and-bad-usage-of-lombok-8c8f70874a93

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -13,50 +15,90 @@
+                break;
+
+            default:
+
```
이곳에도 필요한 것 같군요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -13,50 +15,90 @@
 
 //        GET /tasks
-        if (method.equals("GET") && path.equals("/tasks")) {
+        if (method.equals(GET) && path.equals("/tasks")) {
```
문제는 없을 것 같지만 `"/tasks".equals(method)` 이렇게 바꾸는 것이 좀 더 안전합니다!

왜 그런지 알아보는 것도 재미있는 경험일 것 같아요!

kckoh replied:
혹시 `"/tasks".equals(path)` 말씀 하시는건가요?

ssisksl77 commented:
> > <img alt="error" width="960" src="https://user-images.githubusercontent.com/43446639/190264782-86d03f14-858e-4693-8242-00774816ec9f.png">
> 
> 지금 windows 환경에서 gitbash를 쓰고 있는데 어디서 문제 인지 모르겠네요.

영어로 보낸 것은 문제가 없나요? 

일단 사용하시는 로컬 환경에서 git bash에 한글인코딩을 바꿔보면서 테스트를 해보아야 할 것 같은데요!

일단 영어로 보내는 것이 문제 없다면 한글깨짐 문제는 우선순위를 낮춰도 될 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,147 @@
+        }
+//        GET /tasks/{id}
+        else if (method.equals("GET") && path.contains("/tasks/")) {
+            String[] split = path.split("/");
+
+            String id = split[2];
```
split이라는 변수명이 생성되자마자 바로 밑에서 사용되고 그 다음에는 사용되지 않는 것으로 보여요! `split` 이라는 이름 자체를 없애는 것이 어떨까요?!

kckoh committed:
fix: refactor some code to and implement delete

고친점:
    split[] 변수 x
    default 추가
    allArgs x
    delete API 구현

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -89,23 +100,21 @@ private void methodDispatcher(Method method, String path) throws IOException {
-
 //        GET /tasks
-        if (method.equals(GET) && path.equals("/tasks")) {
+        if (method.equals(GET) && "/tasks".equals(path)) {
```
좋습니다!

ssisksl77 replied:
더해서 `GET.equals(method)` 도 가능할 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -72,12 +72,23 @@ private void methodDispatcher(Method method, String path) throws IOException {
             case DELETE:
+                String id = path.split("/")[2];
+
+                for (Task task : tasks) {
+                    if (task.getId() == Long.parseLong(id)) {
+                        System.out.println(task);
+                        tasks.remove(task);
+                        content = taskToJson(task);
+                        httpStatus = HttpStatus.OK;
+                        return;
+                    }
+                }
```
switch -> for -> if 로 중첩된 코드는 유지보수에 좋지 않습니다! 

가볍게 메소드 추출로 시작해보는 것은 어떨까요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private void sendResponseBody(HttpExchange exchange) throws IOException {
+        OutputStream responseBody = exchange.getResponseBody();
+        responseBody.write(content.getBytes());
+        responseBody.flush();
+        responseBody.close();
```
부득이하게 예외가 발생하는 경우 `close` 가 안될 수 있습니다! 이를 개선해보아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private void addTaskToLists(String body) throws JsonProcessingException {
+        if (!body.isEmpty()) {
```
`null` 체크는 필요없을까요?

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,155 @@
+
+        System.out.println(method + " " + path);
+
+        addTaskToLists(body);
```
body에 값이 있으면 항상 `task` 를 추가하는 메소드로 이해되는데 맞을까요?

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/handler/DemoHttpHandler.java:
```diff
@@ -0,0 +1,155 @@
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    private HttpStatus httpStatus;
+    String content = "Hello World!";
```
`Hello World!` 는 디폴트로 사용되는 값인가요?! 

그렇다면 이름을 변경하고 `final` 키워드도 붙이는 것이 좋을 것 같아요!

ssisksl77 commented:
한주간 회고를 작성하고 링크를 올려주세요!

kckoh commented:
혹시 회고는 영어로 작성해도 될까요? 지금 해외에서 살고 있어서 영어로 작성해서 블로그로 올리고 싶네요!
